### PR TITLE
lightweight alternate implementation of compound models

### PR DESCRIPTION
**Pull request by perrygreenfield**
This module provides an alternate implementation of compound models that
is lighter weight than the default implementation. This is primarily done
to minimize the memory and construction time demands when compound models
become very complex.

The motivation for this arose when JWST NIRSPEC compound models for its
MOS mode were involving almost 100 constituent model elements and on the
order of 300 parameters. Since the default compound model implementation
maps parameters from the consituent components (which themselves are often
compound models), this led to many thousands of parameter instances since`
the total tends to go with the square of the number of model instances.
This led to enormous memory usage and very long instantiation times.
(issue #5985)

The current implementation is already very complex and relies heavily
on metaclass tools making it difficult for many  #developers to understand
and modify. In principle it is possible to optimize the current
implementation but it is very hard to see how to avoid embedding operator
expressions in strings to avoid the many instantiations. Besides being
complex and difficult to add, it adds clumsiness to the user interface
(though admittedly, the alternate solution has its own clumsy aspect).
So this led to a completely different approach that seems much simpler
for certain use cases (it does forgo the ability to fit compound models
for example, though adding such capability without affecting the
performance issues isn't out of the question in the following approach)

This is an initial implementation that implements only very basic
functionality. Some of the functionality that the default implementation
provides will never be provided by this implementation; other functionality
is deferred depending on the experience in using this initial implementation.
Note that this does not affect the underlying basic model classes at all.
This model class does not inherit from the basic model class! (So some
solution to making generic tests for models is needed)

Using this alternate version of compound models requires calling a function
in core to make this one the default. If this is used, **is is highly recommended
that the mode be set back to the default at the end of the code constructing
compound models so that other code depending on the default behavior is
not affected!**

As an example of how to do this:
```python
from astropy.modeling.core import set_compound_model
prevcm = set_compound_model('lite')
compound_model = Gaussian1D(1., 0.5, 0.1) + Gaussian1D(2, 0.7, 0.2)
set_compound_model(prevcm) # the default model type is 'regular'
 ```
Things currently supported:

- evaluation
- inverse evaluation (if possible or provided)

Things not currently supported (but will be if adopted):

- named access to constituent models (And thus their parameters)
- good string representation and such
- conversion to and from the default compound model scheme
- pickling
- possibly more efficient evaluation as done in the style of the current
  compound models (this implementation walks the tree every time)
- and other things I've overlooked at this moment...

Things that will never be supported:

- Compound models of model classes (as opposed to instances)
- Transparent and automatic mapping of parameters of constituent
  models to the compound model level.
- Fitting (the focus of this almost entirely on evaluation)

### CODE REVIEW

**Comment by pllim** [comment]
Can `set_compound_model` be made a context manager? Also, do you have plans to support units for the "lite" version?

**Comment by perrygreenfield** [comment]
@pllim as far as a context manager, I haven't looked into that yet, but perhaps that would be a good way to do it. And yes, unit support would have to be added.

**Comment by nden** [comment]
@perrygreenfield A few questions as I go over this and try it out.

Are the `x` tests included because these are things that will be supported? I am asking because the support for `model.n_submodels()` would be useful. However there are also `x` tests for combining model classes (which you said is not going to be supported) but perhaps this was overlooked.

Parameter change in compound models should still work, right? Even if it's done at the child model level.

Is model indexing going to be supported?

Is combination of lightweight and regular models going to be supported?

So basically this implementation short-circuits the current `ExpressionTree` and other `_CompoundModel` attributes. I think this will require some changes to asdf. More specifically serializing compound models currently walks the expression tree.

These are just first impressions. Please post here any timings or memory estimates if you have available.

**Comment by perrygreenfield** [comment]
@nden the x tests may be supported. I left that as an open issue needing thought. Yes, if you change parameters of constituent models, they should affect the result (though getting at them if you don't already have a reference isn't simple in this version; see next). Indexing of some sort should be provided though whether the user interface will the the same isn't clear (I would have to look at that more closely). But if constituent models (and yes, if some of those constituent models is itself the default compound model type) are named, it will be possible to index by that name (assuming it is unique). If not named, then an alternate scheme, probably the same as the existing one would be used. These references probably would be dug out when requested and then cached for faster access if reused. As for combinations: the alternate can use the original. The original cannot use the alternate (I think that would complicate things quite a bit; and if one wants to use the alternate as part of a original compound model, one simply converts it (through a method or function) as part of the expression. Does this make sense?

**Comment by perrygreenfield** [comment]
@nden As far as an expression tree goes, it does implement one (the evaluation works differently) so  it should be easy to have ASDF walk it as well. As far as n_submodels() goes, I'm not entirely clear from the code documentation, but I gather it is the number of basic models that make up the compound model? If so, that should not be difficult to generate.

Are there other uses of the compound model interface that WCS depends on?

**Comment by nden** [comment]
The `bounding_box` attribute is what comes to mind first.

**Comment by perrygreenfield** [comment]
@nden Presumably the compound model bounding box is an intersection of all constituent model bounding boxes?

**Comment by nden** [comment]
No, it's just a settable property but look at how it's implemented in  [Model.bounding_box](https://github.com/astropy/astropy/blob/master/astropy/modeling/core.py#L916rl).

**Comment by pllim** [comment]
Need to rebase to resolve merge conflict -- FYI. Also I added "experimental" label for now (to fend off astropy-bot), feel free to remove it when it is ready for final review.

**Comment by perrygreenfield** [comment]
This wasn’t intended as a pull request so I’m puzzled as to how it happened.

**Comment by nden** [comment]
Closing this, after offline discussions, in favour of further exploring alternate options.
